frontend/pages/login/[service].tsx

```diff
@@ -23,12 +24,16 @@ export default function AuthReturnPage({ services }) {
   const dispatch = useUserDispatch()
 
   const [pendingTransaction, _setPendingTransaction] = usePendingTransaction()
+  const [returnTo, setReturnTo] = useLocalStorage<string>("returnTo", "")
 
   useEffect(() => {
-    // Redirect to userpage once logged in. Or, if there's a pending transaction, redirect to the purchase page.
+    // Redirect back to where user was, or userpage for unprompted login
     if (user.info && !user.loading) {
       if (pendingTransaction) {
         router.push("/purchase")
+      } else if (returnTo) {
+        router.push(`/${returnTo.replace(".", "/")}`)
```

Comment:
I think this introduces the following security issues:
 - open redirect vulnerability - I can send someone a link to https://beta.flathub.org/login?returnTo=https:..evil.com and this code will redirect them to https://evil.com (which might impersonate the Flathub payment page, for example). See https://cheatsheetseries.owasp.org/cheatsheets/Unvalidated_Redirects_and_Forwards_Cheat_Sheet.html.
 - remote code execution - I can send someone a link to `https://beta.flathub.org/login?returnTo=javascript:alert('Oh no!')` and this code will execute `alert('Oh no!')` in the user's browser.

We should probably validate that we are redirecting them inside the same application. The most straightforward way would probably be to ensure that `returnTo` is always absolute and validate the host:

```javascript
try {
   const parsedReturnTo = new URL(returnTo);

   if (parsedReturnTo.host === location.host) {
      // Safe to redirect
   }
} catch (e) {
  // Show error
}
```

Reply:
Ah, didn't notice the `/` at the beginning (shouldn't review code after midnight :smiling_face_with_tear:). That would prevent the two issues I've described, but resting the security of this code on that one character is probably not ideal (I can easily imagine someone removing it in the future, when the requirements change or during some refactoring, without realizing the security implications).

Reply:
Thanks for flagging this up :smiley:

I agree let's not rely on that `/`, I didn't fully think through the security concerns and intuitively landed on a simple solution which implicitly sweeps them aside. I'll rework this to include more explicit defensive code like you suggest :+1: